hogben/tests/test_fisher.py

```diff
@@ -0,0 +1,337 @@
+    """
+    Create a mock of the refl1d model with a given set of parameters and their
+    bounds
+    """
+    # Parameters described as tuples: (value, lower bound, upper bound)
+    parameter_values = \
+        [(20, 15, 25), (50, 45, 55), (10, 7.5, 8.5), (2, 1.5, 2.5)]
+
+    # Fill parameter values and bounds
+    parameters = [
+        Mock(spec=refnx.analysis.Parameter, value=value,
+             bounds=Mock(lb=lb, ub=ub))
+        for value, lb, ub in parameter_values
+    ]
+    model = Mock(spec=refnx.reflect.ReflectModel, xi=parameters)
+    model.xi = parameters
+    return model
+
+
+@pytest.fixture
+def mock_refl1d_model():
+    """
+    Create a mock of the refl1d model with a given set of parameters and their
+    bounds
+    """
+    # Parameters described as tuples: (value, lower bound, upper bound)
+    parameter_values = \
+        [(20, 15, 25), (50, 45, 55), (10, 7.5, 8.5), (2, 1.5, 2.5)]
+
+    # Fill parameter values and bounds
+    parameters = [
+        Mock(spec=bumps.parameter.Parameter, value=value,
+             bounds=Mock(limits=[lb, ub]))
+        for value, lb, ub in parameter_values
+    ]
+    model = Mock(spec=refl1d.experiment.Experiment, xi=parameters)
+    model.xi = parameters
+    return model
+
+
+def get_fisher_information(models: list[Union['refnx.reflect.ReflectModel',
+                                              'refl1d.experiment.Experiment']],
+                           xi: Optional[
+                           list[Union['refnx.analysis.Parameter',
+                                      'bumps.parameter.Parameter']]] = None,
+                           counts: Optional[list[int]] = None,
+                           qs: Optional[list[list[float]]] = None,
+                           step: float = 0.005) -> np.ndarray:
+    """
+    Obtains the Fisher matrix given a specified model, the optional
+    parameters are given a default value if not specified.
+    Args:
+        models: List of models that describe the experiment.
+        xi: The varying model parameters.
+        counts: incident neutron counts corresponding to each Q value.
+        qs: The Q points for each model.
+        step: step size to take when calculating gradient.
+    Returns:
+        numpy.ndarray: Fisher information matrix for given models and data.
+    """
+    # Provide default values for qs, counts and xi
+    if qs is None:
+        # Add one list of qs per model
+        qs = [np.array([0.1, 0.2, 0.4, 0.6, 0.8]) for _ in models]
+    if counts is None:
+        # Define the incident counts at each q at 100 counts per point
+        count_list = np.ones(len(qs[0])) * 100
+        counts = [count_list for _ in models]  # One list of counts per model
+    if xi is None:
+        xi = []  # Concatenate all xi's to a single list of parameters.
+        for model in models:
+            xi += model.xi
+    return fisher(qs, xi, counts, models, step)
+
+
+def get_reflectivity_given_datapoints(data_points):
```

Comment:
Unfortunately this does not work directly. The `side_effect` attribute needs to be called on a generator function, which `np.random.rand(len(qs))` is not. I can set the `return_value` using `mock_reflectivity.return_value = np.random.rand(len(qs))`, but that executes the function only once and then returns the same value on subsequent calls.

A thing I can do, is to simplify `get_reflectivity_given_datapoints` significantly using the suggested comment here:
```
def get_random_reflectivity(data_points):
    while True:
        yield np.random.rand(data_points)
```

Which yields a different array of reflectivity points each time, and indeed passes testing because it doesn't actually use the exact values of the returned numbers.

Reply:
I'm pretty sure the generator doesn't need to be its own function

Reply:
You can indeed define a generator function in a list-comprehension style using `()` brackets, I was not aware of that.
Do mind that a generator function slowly gets emptied, and you cannot create an infinite loop. 

The suggested function just yields the random values from a single array (generated by `np.random.rand(len(qs))`), instead of yielding the list itself. But what does work is `mock_reflectivity.side_effect = (np.random.rand(len(qs)) for _ in range(9))`. I'm not sure if I'm the biggest fan of the unused index to let it loop a predefined amount of 9 times, but I couldn't find a clean solution here.

The 9 in this range is because reflectivity is called eight times. Twice per time the unit test is performed. I guess it should work with 8 as well, but it doesn't, perhaps it yields the first value in the generator function on assignment.